reporting/impl/src/main/java/org/jboss/windup/reporting/rules/rendering/RenderReportRuleProvider.java

```diff
@@ -54,60 +63,100 @@ public RenderReportRuleProvider()
-                {
+                public void perform(GraphRewrite event, EvaluationContext context, ReportResourceFileModel payload) {
                     ReportService reportService = new ReportService(event.getGraphContext());
                     Path outputDir = Paths.get(reportService.getReportDirectory());
 
                     File directory = outputDir.toFile();
                     File fullPath = new File(directory, FilenameUtils.separatorsToSystem("resources/" + payload.getPrettyPath()));
 
-                    try
-                    {
+                    try {
                         FileUtils.forceMkdir(fullPath.getParentFile());
                         FileUtils.copyFile(payload.asFile(), fullPath);
                         LOG.info("Copied raw file: " + payload.getFilePath() + " to: " + fullPath.getAbsolutePath());
-                    }
-                    catch (IOException e)
-                    {
+                    } catch (IOException e) {
                         LOG.warning("Exception creating file: " + fullPath.getAbsolutePath());
                     }
                 }
             });
     }
     // @formatter:on
+
+    private class FreeMarkerThreadedRenderer extends GraphOperation
+    {
+        private final Furnace furnace;
+
+        public FreeMarkerThreadedRenderer(Furnace furnace)
+        {
+            this.furnace = furnace;
+        }
+
+        @Override
+        public void perform(final GraphRewrite event, final EvaluationContext context)
+        {
+            Iterable<? extends WindupVertexFrame> reportModelsIterable = Variables.instance(event)
+                        .findVariable(Iteration.DEFAULT_VARIABLE_LIST_STRING);
+            final Queue<WindupVertexFrame> reportModels = new LinkedList<>();
+            for (WindupVertexFrame frame : reportModelsIterable)
+                reportModels.add(frame);
+            final FreeMarkerIterationOperation freeMarkerIterationOperation = FreeMarkerIterationOperation.create(furnace);
+            final IterationProgress iterationProgress = IterationProgress.monitoring("Rendering Reports", 100);
+
+            int threadCount = WindupExecutors.getDefaultThreadCount();
+            ExecutorService executorService = WindupExecutors.newFixedThreadPool(WindupExecutors.getDefaultThreadCount());
+
+            // Set the frames as iteration progress uses this
+            event.getRewriteContext().put(Iteration.DEFAULT_VARIABLE_LIST_STRING, reportModels);
+            iterationProgress.perform(event, context);
+
+            for (int i = 0; i < threadCount; i++)
+            {
+                executorService.submit(new Callable<Void>()
+                {
+                    @Override
+                    public Void call() throws Exception
+                    {
+                        while (true)
+                        {
+                            final ReportModel reportModel;
+                            synchronized (reportModels)
```

Comment:
I'm not sure that a blockingqueue would add much here vs just building the work queue ahead of time.

Having said that, this should be using a concurrentlinkeddeque instead of all the extra synchronization. Will fix...